### ISSUE TITLE
Create telemetry broadcast channel

### ISSUE THREAD

**Issue by hekonsek** [rhiot/rhiot#467]
@ppatierno :
> Regarding the hot path always on telemetry pattern, I think 
> that the RPC style with a strong binding with a related "bean" means strongly typing the 
> scenario : I see that devices and related data are strongly coupled with the service 
> which "eat" (process) the telemetry data.
> Even in that case telemetry data should be readable from different backend services, in 
> that case it means calling different beans today.

We can create a multicast channel for telemetry data, where document service will be just on of the consumer of the messages coming to this service. This way we can have all the benefits of storing telemetry in document service (like CRUD REST API for collected data), but you can also consume data using any other consumer you want.

**Comment by ppatierno** [comment]
A multicast is needed for sure to have more readers on telemetry data but I have the opinion that the "ingestion" stream should decouple sender (devices) from receiver (backend service).
With current type of service binding and calling bean methods, producers and consumer are strongly tight. What happens if the service isn't available ?
In a telemetry scenario, it's useful to have events sent to a stream a devices pace and read from backend services at their own pace. Of course a device is interested to know if its message was delivered to the IoT system (but not to the backend service).
However, all depends on what should be the experience to give a customer to build an IoT solution on this platform : if the customer is limited to use this kind of service binding and bean calls or the customer should have its own backend service that can read telemetry data from the platform using a standard protocol (for example AMQP).

**Comment by hekonsek** [comment]
Actually device and service are currently not coupled, as broker will cache the messages until service is available again. :)

Keep in mind that document service is not THE service to store telemetry data. It can be used for that purpose, if you are interested in having CRUD API for your data available out of the box. But it is just one of the service interested in consuming that stream.

If we would have telemetry broadcast topic, then we (or end-user) could create other services interested in consuming telemetry data. Device service could be one of them. Spark Streaming API could be the other. Kafka Stream could be yet another. Custom service created by our customers is just another.

So my proposition would be to have a convention that some AMQP destinations (for example starting with `telemetry.*` are broadcast channels that can be used by other services to listen for telemetry data. It is up to the end user to choose the services she is interested in to wire it to that channel. For example I could be interested in Document Service and Spark Streaming API service consuming from this channel. But customer A could be interested in Spark Streaming and Kafka services listening to this only. 

What do you think? :)

**Comment by hekonsek** [comment]
The important thing is that I would not assume stream persistence by default. It should be up to the end use to decide if they want to have stream persisted by given channel listener. For example many users would like to use Spark Streaming API processing only. All in-flight without persistence of the telemetry.

**Comment by ppatierno** [comment]
Regarding the decoupling you are right ... I lost one point : the service has a camel route to read from amqp and call the bean ... if the service is offline, the entire route is offline, no one reads from amqp so messages are still there :-)

Regarding telemetry path ... I'm thinking about companies that don't want to develop a service following our interfaces but they could have a service which reads telemetry messages using an AMQP client inside their application/backend layer.
The telemetry channel could be only one but in the current architecture there is always the operation in the path (at least, with no parameters). It could be difficult for an external service (not using our interface to listen on all possible channels), isn't it ?

Regarding using ActiveMQ as ingestion system on the telemetry pipeline I have some concerns about performance. It seems that Kafka could be better than ActiveMQ but I already saw that Kafka uses a custom protocol and not a standard one (AMQP, MQTT, ...) ... so it's bad for me ! :-)

**Comment by hekonsek** [comment]
> The telemetry channel could be only one but in the current architecture there is always 
> the operation in the path (at least, with no parameters). It could be difficult for an external
> service (not using our interface to listen on all possible channels), isn't it ?

I would say that we a broadcast channels with raw telemetry data. All broadcast channels should follow some naming convention. For example starts with `/telemetry/`AMQP path. Like `/telemetry/temperature` or `/telemetry/foo`. Device willing to sent telemetry data just sends the payloads to the channel they want (for example `/telemtry/baz`). Now backend services willing to read telemetry stream of data can just subscribe to the selected channel (for example `/telemetry/baz`). You can also subscribe your backend services to consume from all the telemetry channels using AMQP wildcards (`/telemetry/*`).

You can also OPTIONALLY uses our service binding to bind Java service that AMQP telemetry stream. But this is optional. However if I you develop Java service, then using our service binding is a wise decision, as it reduces the glue code you need to write.

Also you could be able to configure `DocumentService` to connect to selected telemetry broadcast stream. For example I'd like to have CRUD API over telemetry data from `/telemtry/baz`. So I connect `DocumentService` to consume telemetry data from that stream. But only if you want to :) .

What do you think? Maybe we could have a Skype call to talk about it?

>Regarding using ActiveMQ as ingestion system on the telemetry pipeline I have some 
> concerns about performance. It seems that Kafka could be better than ActiveMQ but I 
> already saw that Kafka uses a custom protocol and not a standard one 
> (AMQP, MQTT, ...) ... so it's bad for me ! :-)

Keep in mind that we don't use ActiveMQ. Use we AMQP, which can be brokerless. In particular you can created Qpid Dispatch Router [1] based network of routers which can handle a tons of concurrent connection and then just connect to that stream of data with anything you need. Kafka, Spark, DocumentService, ElasticSearch, custom services, etc. You can also use Qpid Dispatch Router to form a cluster of Artemis brokers hidden behind a network of routers.

So I wouldn't be concerned about scalability here :) .

[1] http://qpid.apache.org/components/dispatch-router/

**Comment by ppatierno** [comment]
I'm always concerned about channels explosion :-) 
Having the "destination" (ex. temp, foo, ...) in the path can open to million/billion/trillion of possibilities. 
I think that in the /telemetry path only one information is useful to have and it's the sender id (ex. the device id).
Using the headers/properties could be a good solution for all other information.
I think that on this point we agree (regarding to remove info on the URI path).

Of course I agree with you regarding no problem with scalability on cloud side ... but I'm thinking about low-resources constrained devices too. I'm quite sure that a lot of developers will use more than one TCP connection for different paths (ex. /telemetry/temp, /telemetry/foo) and not only one with multiplexed links inside it. It's another concern ...

The other side is that removing more info from the path we can't use it as a topic on MQTT protocol but you are saying that we won't use ActiveMQ with MQTT native support, so we'll need an adapter to transform message from MQTT to AMQP. Is it your idea ?

BTW ... a Skype call is needed :-)

**Comment by hekonsek** [comment]
Yeah, we can remove the name of the telemetry name from path. And leave only `/telemetry`. The name can be sent in header or body of the message. That should spare some extra topics on a broker side :) .

Looking forward for our call! :)
